tsdb/importer/importer.go

```diff
@@ -0,0 +1,602 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importer
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/gob"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/pkg/value"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Implementing the error interface to create a
+// constant, which cannot be overridden.
+// https://dave.cheney.net/2016/04/07/constant-errors
+type Error string
+
+func (e Error) Error() string {
+	return string(e)
+}
+
+const (
+	// This error is thrown when the current entry does not have a timestamp associated.
+	NoTimestampError = Error("expected timestamp with metric")
+	// This error is thrown when the sample being parsed currently has a corresponding block
+	// in the database, but has no metadata collected for it, curiously enough.
+	NoBlockMetaFoundError = Error("no metadata found for current samples' block")
+)
+
+type blockTimestampPair struct {
+	start, end int64
+}
+
+type newBlockMeta struct {
+	index      int
+	count      int
+	mint, maxt int64
+	isAligned  bool
+	dir        string
+	children   []*newBlockMeta
+}
+
+// Content Type for the Open Metrics Parser.
+// Needed to init the text parser.
+const contentType = "application/openmetrics-text;"
+
+// ImportFromFile imports data from a file formatted according to the Open Metrics format,
+// converts it into block(s), and places the newly created block(s) in the
+// TSDB DB directory, where it is treated like any other block.
+func ImportFromFile(
+	r io.Reader,
+	dbPath string,
+	maxSamplesInMemory int,
+	maxBlockChildren int,
+	logger log.Logger,
+) error {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	level.Debug(logger).Log("msg", "creating temp directory")
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	level.Debug(logger).Log("msg", "reading existing block metadata to infer time ranges")
+	blockMetas, err := constructBlockMetadata(dbPath)
+	if err != nil {
+		return err
+	}
+
+	level.Info(logger).Log("msg", "importing input data and aligning with existing DB")
+	if err = writeSamples(r, tmpDbDir, blockMetas, maxSamplesInMemory, maxBlockChildren, logger); err != nil {
+		return err
+	}
+
+	level.Info(logger).Log("msg", "merging fully overlapping blocks")
+	if err = mergeBlocks(tmpDbDir, blockMetas, logger); err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "copying newly created blocks from temp location to actual DB location")
```

Comment:
What if location is remote?

Reply:
I haven't tested that specific case, but that seems like something the OS / golang libs itself should handle right?

Reply:
How? The key is to not overcomplicate the program. I would suggest simple thing. Just `output` dir and nothing more. (: